File: backend/internal/server/routes/oauth.go
package routes

import (
	"github.com/Wei-Shaw/sub2api/internal/handler"
	"github.com/Wei-Shaw/sub2api/internal/server/middleware"

	"github.com/gin-gonic/gin"
)

// RegisterOAuthWellKnown mounts unauthenticated discovery endpoints at the
// server root (NOT under /api) — JWKS must live at the well-known path.
func RegisterOAuthWellKnown(r gin.IRouter, h *handler.OAuthHandler) {
	r.GET("/.well-known/jwks.json", h.JWKS)
}

// RegisterOAuthAPIRoutes mounts the authenticated OAuth self-service surface
// under /api/oauth — deliberately NOT /api/v1, since this path is shared
// with the unauthenticated OAuth endpoints (device/code, token, etc. — later
// tasks in this plan) and is consumed by the hermes CLI / dashboard, not the
// panel's versioned API.
func RegisterOAuthAPIRoutes(r gin.IRouter, h *handler.OAuthHandler, jwtAuth middleware.JWTAuthMiddleware) {
	authenticated := r.Group("/api/oauth")
	authenticated.Use(gin.HandlerFunc(jwtAuth))
	{
		authenticated.POST("/self-hosted-client", h.RegisterSelfHostedClient)
		// Device approval screen (Task 7): a logged-in human confirming or
		// rejecting a device-flow login in inferno-frontend's browser UI.
		// Session-authenticated like self-hosted-client above, unlike
		// device/code and token below (RegisterOAuthDeviceRoutes) which have
		// no session by design.
		// What the human is being asked to approve. Session-gated for the same
		// reason approve/deny are, and additionally because an unauthenticated
		// version would let anyone holding a phished user_code confirm it is
		// live and read back the scopes it carries.
		authenticated.GET("/device/pending", h.PendingDeviceAuthorization)
		authenticated.POST("/device/approve", h.ApproveDevice)
		authenticated.POST("/device/deny", h.DenyDevice)
		// AuthorizeConsentView.vue's display-name lookup for the
		// /oauth/authorize consent screen (Task 4). See
		// OAuthHandler.PendingAuthorization's doc comment for why this is
		// on the envelope while /oauth/authorize itself is bare.
		authenticated.GET("/authorize/pending", h.PendingAuthorization)
	}
}

// RegisterOAuthDeviceRoutes mounts the UNAUTHENTICATED /api/oauth device-flow
// endpoints. Deliberately a separate group from RegisterOAuthAPIRoutes (which
// requires jwtAuth): a headless CLI calling POST /api/oauth/device/code has
// no session yet — client_id is the identity for this request, not a bearer
// token. Both groups share the /api/oauth prefix; gin allows two distinct
// *gin.RouterGroup values rooted at the same path as long as their route
// patterns don't collide.
//
// POST /api/oauth/token belongs here for the same reason: it is the
// credential-issuance step itself (device_code and refresh_token grants) —
// the caller by definition has no session/bearer token yet, so it cannot be
// gated behind jwtAuth.
func RegisterOAuthDeviceRoutes(r gin.IRouter, h *handler.OAuthHandler) {
	unauthenticated := r.Group("/api/oauth")
	{
		unauthenticated.POST("/device/code", h.DeviceCode)
		unauthenticated.POST("/token", h.Token)
	}
}

// RegisterOAuthAuthorizeRoute mounts GET/POST /oauth/authorize (Task 4) at
// the server ROOT, not under /api -- it is a browser endpoint, reached by
// the hermes client's system browser via a real top-level navigation, not
// an XHR call, so it cannot live behind the header-only jwtAuth group that
// RegisterOAuthAPIRoutes uses (a raw navigation never carries an
// Authorization header at all; jwtAuth would 401 every unauthenticated hit
// with a JSON body no browser can act on).
//
// optionalJWTAuth (middleware.OptionalJWTAuthMiddleware, already wired for
// every other panel endpoint) is used instead: no header at all passes
// through anonymously (the handler then does its own 302-to-login), a
// header on a still-valid session resolves the AuthSubject the handler
// needs for the ownership check, and a header on an EXPIRED/invalid session
// is rejected by that shared middleware before the handler runs -- see
// OAuthHandler.Authorize's doc comment for why that one case is this
// otherwise-bare endpoint's sole JSON exception.
func RegisterOAuthAuthorizeRoute(r gin.IRouter, h *handler.OAuthHandler, optionalJWTAuth middleware.OptionalJWTAuthMiddleware) {
	r.GET("/oauth/authorize", gin.HandlerFunc(optionalJWTAuth), h.Authorize)
	r.POST("/oauth/authorize", gin.HandlerFunc(optionalJWTAuth), h.Authorize)
}

// RegisterOAuthAccountRoutes mounts GET /api/oauth/account behind Task 6's
// middleware.RequireOAuthScope — a THIRD distinct /api/oauth group,
// deliberately not merged into RegisterOAuthAPIRoutes (gated by jwtAuth,
// Inferno's HMAC panel-session middleware) or RegisterOAuthDeviceRoutes
// (unauthenticated). A hermes agent calling this endpoint presents an
// OAuth-issued ES256 bearer token, not a panel session and not a bare
// client_id — a third auth shape needs its own group. required scope is ""
// (any validly-signed, unexpired OAuth token) since this endpoint only
// resolves which org(s) the token's holder belongs to; it grants no
// elevated capability itself.
func RegisterOAuthAccountRoutes(r gin.IRouter, h *handler.OAuthHandler) {
	scoped := r.Group("/api/oauth")
	{
		scoped.GET("/account", middleware.RequireOAuthScope(h.KeyService(), ""), h.Account)
	}
}

File: inferno-frontend/src/api/oauth.ts
/**
 * OAuth device-flow approval API.
 *
 * ApproveDevice/DenyDevice (backend/internal/handler/oauth_handler.go) are
 * panel-only endpoints consumed exclusively by DeviceApprovalView -- unlike
 * every other /api/oauth endpoint, they are session-authenticated and wrapped
 * in internal/pkg/response's {code,message,data} envelope, which apiClient's
 * response interceptor already unwraps/normalizes the same way it does for
 * the rest of the panel API. They live under /api/oauth, not /api/v1
 * (apiClient's baseURL), so the request targets an absolute URL built by
 * buildGatewayUrl rather than a bare relative path.
 */
import { apiClient, buildGatewayUrl } from './client'

export interface DeviceDecisionResult {
  status: 'approved' | 'denied'
}

export async function approveDevice(userCode: string): Promise<DeviceDecisionResult> {
  const { data } = await apiClient.post<DeviceDecisionResult>(
    buildGatewayUrl('/api/oauth/device/approve'),
    { user_code: userCode }
  )
  return data
}

export async function denyDevice(userCode: string): Promise<DeviceDecisionResult> {
  const { data } = await apiClient.post<DeviceDecisionResult>(
    buildGatewayUrl('/api/oauth/device/deny'),
    { user_code: userCode }
  )
  return data
}

/**
 * What the human is being asked to approve, resolved from the user_code they
 * typed or that verification_uri_complete prefilled.
 *
 * RFC 8628 section 5.4 requires the authorization server to display client and
 * authorization information at the verification URI. Without it the approval
 * screen was a bare code box and two buttons -- indistinguishable from an
 * ordinary login prompt, which is exactly what makes a phished user_code
 * dangerous: the victim approves without ever seeing who is asking or what
 * they are handing over.
 *
 * Session-authenticated like approveDevice/denyDevice, so an unauthenticated
 * holder of a code cannot use it to confirm the code is live.
 */
export interface PendingDeviceAuthorization {
  client_name: string
  client_id: string
  scopes: string[]
  expires_at: string
}

export async function fetchPendingDeviceAuthorization(
  userCode: string
): Promise<PendingDeviceAuthorization> {
  const { data } = await apiClient.get<PendingDeviceAuthorization>(
    buildGatewayUrl('/api/oauth/device/pending'),
    { params: { user_code: userCode } }
  )
  return data
}

/**
 * GET/POST /oauth/authorize (backend/internal/handler/oauth_authorize_handler.go)
 * -- the RFC 6749 §4.1 authorization_code + PKCE browser leg.
 *
 * DELIBERATELY NOT on the {code,message,data} envelope, and NOT JSON at all:
 * that endpoint is a real OAuth authorization endpoint any client can
 * navigate a browser to directly, so its success/redirect-with-error
 * responses are a bare `text/plain` body carrying the exact URL to navigate
 * to next (the browser cannot follow a literal cross-origin 302 issued to a
 * `fetch()` call the way it could a real top-level navigation -- see that
 * handler's doc comment for the full reasoning). AuthorizeConsentView.vue's
 * only job on a "redirect" result is `window.location.assign(url)`.
 *
 * checkAuthorize is the GET: it reports either "here is where to go next"
 * (auto-approved, or the request was rejected with an RFC redirect-class
 * error) or "202, show the consent screen". decideAuthorize is the
 * consent screen's explicit Approve/Deny action (POST); both always resolve
 * to a target URL.
 */
export interface AuthorizeParams {
  response_type: string
  client_id: string
  redirect_uri: string
  scope: string
  state: string
  code_challenge: string
  code_challenge_method: string
}

export type AuthorizeCheckResult =
  | { kind: 'redirect'; url: string }
  | { kind: 'consent_required' }

export async function checkAuthorize(params: AuthorizeParams): Promise<AuthorizeCheckResult> {
  const response = await apiClient.get<string>(buildGatewayUrl('/oauth/authorize'), {
    params,
    responseType: 'text',
  })
  if (response.status === 202) {
    return { kind: 'consent_required' }
  }
  return { kind: 'redirect', url: response.data }
}

export async function decideAuthorize(
  params: AuthorizeParams,
  decision: 'approve' | 'deny'
): Promise<string> {
  const response = await apiClient.post<string>(
    buildGatewayUrl('/oauth/authorize'),
    null,
    { params: { ...params, decision }, responseType: 'text' }
  )
  return response.data
}

/**
 * Display-only "who is asking" lookup for the consent screen -- unlike
 * checkAuthorize/decideAuthorize above, this IS on the panel envelope (see
 * OAuthHandler.PendingAuthorization's doc comment).
 */
export interface PendingAuthorization {
  client_name: string
  client_id: string
}

export async function fetchPendingAuthorization(clientID: string): Promise<PendingAuthorization> {
  const { data } = await apiClient.get<PendingAuthorization>(
    buildGatewayUrl('/api/oauth/authorize/pending'),
    { params: { client_id: clientID } }
  )
  return data
}
